### ISSUE TITLE
pdf dashed lines trigger mis-rendering by ghostscript

### ISSUE THREAD

**Issue by anntzer** [matplotlib/matplotlib#10036]
### Bug report

**Bug summary**

The pdf backend renders dashed lines in a way that appears to trigger a rendering bug with ghostscript.

In the following image, the top row is the test baseline scaled_lines.pdf (for test_lw_scaling) rendered by okular 17.08.3 and gs 9.22 (the latter is used by the test machinery to rasterize images before image comparison)  It is clear that gs renders some dahses with a partially "rounded" edge.

The bottom row is the same test images generated by mplcairo's pdf output: here, gs renders the dashes correctly.

![scaled_lines](https://user-images.githubusercontent.com/1322974/34086442-aa50ace0-e350-11e7-8a63-5b1ed30f95c7.png)

While this may be dismissed as an upstream bug, it may be worth investigating how difficult it is to generate a pdf that gs renders correctly.

**Matplotlib version**

  * Operating system:
  * Matplotlib version: master
  * Matplotlib backend (`print(matplotlib.get_backend())`):
  * Python version:
  * Jupyter version (if applicable):
  * Other libraries:

**Comment by tacaswell** [comment]
Is the diff between the pdfs meaningful?

**Comment by anntzer** [comment]
A minimal example triggering the issue is
```
import matplotlib.pyplot as plt
plt.rcdefaults()
fig, ax = plt.subplots()
ax.set_axis_off()
ax.plot([0] * 6, ":", lw=10)
fig.savefig("/tmp/test.pdf")
plt.show()
```
In particular, it seems to only trigger when there are multiple "aligned" line segments.

The resulting pdfs (for the minimal example) are attached.
(matplotlib)
[test-pdf.pdf](https://github.com/matplotlib/matplotlib/files/1576498/test-pdf.pdf)
(mplcairo)
[test-cairo.pdf](https://github.com/matplotlib/matplotlib/files/1576500/test-cairo.pdf)
I had a look at the raw streams but they are basically completely different and I don't know the pdf format well enough (or at all...) to figure out where the relevant differences are.

**Comment by mpetroff** [comment]
TLDR: I'm fairly certain this is due to a bug in Ghostscript, but we may still want to produce PDF 1.5 instead of PDF 1.4.

Here's the relevant portion of the deflated drawing stream from the PDF backend with annotations:
```
[10 16.5] 0 d                   # Set the line dash pattern to 10 on, 16.5 off
1 j                             # Set the line join style to rounded
10 w                            # Set the line width to 10
73.832729 171.072 m             # Move to x y (no draw)
138.76364 171.072 l             # Draw to x y
203.69455 171.072 l             # Draw to x y
268.62547 171.072 l             # Draw to x y
333.55638 171.072 l             # Draw to x y
398.48728 171.072 l             # Draw to x y
.12156863 .46666668 .7058824 RG # Set the stroking color to blue
S                               # Stroke the path
```

Here's the relevant portion from the Cairo backend:
```
[10 16.5] 0 d                   # Set the line dash pattern to 10 on, 16.5 off
1 j                             # Set the line join style to rounded
10 w                            # Set the line width to 10
10 M                            # Set the miter limit to 10
73.832 174.527 m                # Move pen to x y (no draw)
398.488 174.527 l               # Draw to x y
.121569 .466667 .705882 RG      # Set the stoke color to blue
S                               # Stroke the path
```

The relevant difference is the PDF backend draws the line in multiple steps, but the Cairo backend does it all at once.

The Cairo backend also sets the miter limit, but this is irrelevant since the line join style is set to `rounded`, not `miter`. It uses a slightly different Y coordinate due to a previous coordinate transform, but this is also not relevant.

Removing the intermediate draw commands from the PDF backend output causes the rendering in Ghostscript to match that of the Cairo backend output.

Per the [PDF 1.7 spec](https://opensource.adobe.com/dc-acrobat-sdk-docs/standards/pdfstandards/pdf/PDF32000_2008.pdf), section 8.4.3.4:
> Join styles shall be significant only at points where consecutive segments of a path connect at an angle; segments that meet or intersect fortuitously shall receive no special treatment.

Per my reading of this part of the spec, the join style should not be applied because the line segments are straight. Thus, I'd consider this to be a bug in Ghostscript.

However, we're producing [PDF 1.4](https://ipgdoc.epfl.ch/lib/exe/fetch.php?media=epfl:software_guides:pdfreference.pdf), which has the following caveat about round joins in Table 4.5 of the spec:
> Note: If path segments shorter than half the line width meet at a sharp angle, an unintended “wrong side” of the circle may appear.

The PDF spec was changed in version 1.5 to address this issue. Thus, I suspect Ghostscript is using an implementation for round joins based on the method in pre-1.5 versions of the PDF spec. Even if I change the PDF version to 1.5, the issue in Ghostscript does not go away, so I suspect Ghostscript uses a single rendering method for round joins regardless of the version of the PDF being rendered and that it is based on the pre-1.5 method. However, we might want to consider bumping the PDF output version to version 1.5, since a renderer that implements the PDF spec to the letter would produce rendering artifacts similar to the ones here in certain cases.

**Comment by mpetroff** [comment]
We could workaround the Ghostscript bug in cases with straight lines by optimizing the drawing operation to draw the line all at once, as is done by the Cairo backend.

**Comment by anntzer** [comment]
Thanks for the investigation.  Do you think this is something you could report to ghostscript?

**Comment by mpetroff** [comment]
I created a minimal test case and reported the bug upstream: https://bugs.ghostscript.com/show_bug.cgi?id=705653

**Comment by anntzer** [comment]
Thanks :)

**Comment by mpetroff** [comment]
This was just fixed upstream in Ghostscript: https://git.ghostscript.com/?p=ghostpdl.git;h=04b799c66a8993af13c4b9b5fafb3dcfa44a16c5

**Comment by anntzer** [comment]
Thanks again for your investigation.  Let's close this; if someone wants to integrate path simplification with the pdf backend that may still be a good idea (although I don't know how much slowdown that would cost) but I don't think it's worth keeping this open to track that.
